lib/throttle-stream.js

```diff
@@ -0,0 +1,153 @@
+'use strict'
+
+const { Transform } = require('stream')
+
+/**
+ * The `ThrottleStream` is very similar to the node core
+ * `stream.Transform` stream, except that you specify a `bps` "bytes per
+ * second" option and data *will not* be passed through faster than the byte
+ * value you specify.
+ *
+ * You can invoke it with a function for `bps` that takes the elapsed time (in seconds)
+ * and the number of bytes already sent as parameters. The function should return
+ * the desired bytes per second value.
+ *
+ * @param {Object} opts an options object or the "bps" function value
+ * @api public
+ * @constructor ThrottleStream
+ * @extends Transform
+ * @param {Object} [opts] - options
+ * @param {number|Function} [opts.bps] - number or function that returns bytes per second to throttle to
+ */
+function ThrottleStream (opts) {
+  if (!new.target) {
+    return new ThrottleStream(opts)
+  }
+
+  Transform.call(this, opts)
+
+  /**
+   * The function to calculate bytes per second to throttle to
+   * @type {Function}
+   * @public
+   * @memberof ThrottleStream
+   * @name bpsFn
+   * @default null
+   */
+  this.bpsFn = typeof opts?.bps === 'undefined'
+    ? function bpsFn (_elapsedTime, _bytes) {
+      return 16384
+    }
+    : typeof opts.bps === 'number'
+      ? function bpsFn (_elapsedTime, _bytes) {
+        return opts.bps
+      }
+      : opts.bps
+
+  /**
+   * The time that the stream started
+   * @type {number}
+   * @public
+   * @memberof ThrottleStream
+   * @name startTime
+   * @default null
+   */
+  this.startTime = null
+
+  /**
+   * The number of bytes that have been sent through the stream
+   * @type {number}
+   * @public
+   * @memberof ThrottleStream
+   * @name bytes
+   * @default 0
+   */
+  this.bytes = 0
+
+  /**
+   * The buffered chunk.
+   * @type {Buffer}
+   * @private
+   * @memberof ThrottleStream
+   * @name _buffer
+   * @default null
+   */
+  this._buffer = null
+
+  this._interval = null
+
+  /**
+   * Number of bytes allowed to be pushed to the next stream
+   * @type {number}
+   * @private
+   * @default 0
+   */
+  this._allowedBytes = 0
+
+  this._callback = null
+}
+
+ThrottleStream.prototype = Object.create(Transform.prototype, {
```

Comment:
same comment about maybe using a class instead

Reply:
Idk. @mcollina is not a fan of the ES6 Classes for streams iirc

Reply:
Isn't Readable literally a class?

Reply:
I will ask @mcollina regarding this.

Reply:
The only difference between the two is that `ThrottleStream()` would work with this and not with `class`. Use whatever style you prefer.

Reply:
Tbh i love the old school prototypical inheritance way.